homeassistant/components/obihai/sensor.py

```diff
@@ -44,43 +44,41 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
 
     sensors = []
 
-    pyobihai = PyObihai()
+    pyobihai = PyObihai(host, username, password)
 
-    login = pyobihai.check_account(host, username, password)
+    login = pyobihai.check_account()
     if not login:
         _LOGGER.error("Invalid credentials")
         return
 
-    services = pyobihai.get_state(host, username, password)
+    services = pyobihai.get_state()
 
-    line_services = pyobihai.get_line_state(host, username, password)
+    line_services = pyobihai.get_line_state()
 
-    call_direction = pyobihai.get_call_direction(host, username, password)
+    call_direction = pyobihai.get_call_direction()
 
     for key in services:
-        sensors.append(ObihaiServiceSensors(pyobihai, host, username, password, key))
+        sensors.append(ObihaiServiceSensors(pyobihai, key))
 
     for key in line_services:
-        sensors.append(ObihaiServiceSensors(pyobihai, host, username, password, key))
+        sensors.append(ObihaiServiceSensors(pyobihai, key))
 
     for key in call_direction:
-        sensors.append(ObihaiServiceSensors(pyobihai, host, username, password, key))
+        sensors.append(ObihaiServiceSensors(pyobihai, key))
 
     add_entities(sensors)
 
 
 class ObihaiServiceSensors(Entity):
     """Get the status of each Obihai Lines."""
 
-    def __init__(self, pyobihai, host, username, password, service_name):
+    def __init__(self, pyobihai, service_name):
         """Initialize monitor sensor."""
-        self._host = host
-        self._username = username
-        self._password = password
         self._service_name = service_name
         self._state = None
         self._name = f"{OBIHAI} {self._service_name}"
         self._pyobihai = pyobihai
+        self._unique_id = f"{self._pyobihai.get_device_serial()}-{self._service_name}"
```

Comment:
Does this do an api call with I/O or just return a cached serial number? We want to avoid side effects in init method.

Reply:
This will reach out to the device to grab the serial number.

Reply:
Should I move the calls back to the property? This is how I had it before and thought it was wrong :)

https://github.com/home-assistant/home-assistant/pull/26922/commits/3e16401007fcecff8e3a8ab385a768793a120593#diff-e21c7150c6e210babbc267e879addf34R102-R104

Reply:
Fetch the serial number before creating the entity and pass the number as an argument to the entity.

Reply:
done :)